graphbolt/src/cnumpy.cc

```diff
@@ -88,145 +97,159 @@ void OnDiskNpyArray::ParseNumpyHeader(torch::Tensor shape) {
-                          &io_uring_queue_[thread_id], &complete_queue) < 0) {
-                    perror("io_uring_wait_cqe");
-                    abort();
-                  }
-                  struct io_uring_cqe *complete_queues[group_size_];
-                  int cqe_count = io_uring_peek_batch_cqe(
-                      &io_uring_queue_[thread_id], complete_queues,
-                      group_size_);
-                  if (cqe_count == -1) {
-                    perror("io_uring_peek_batch error\n");
-                    abort();
-                  }
-                  // Move the head pointer of completion queue.
-                  io_uring_cq_advance(&io_uring_queue_[thread_id], cqe_count);
-                  num_finish += cqe_count;
+c10::intrusive_ptr<Future<torch::Tensor>> OnDiskNpyArray::IndexSelectIOUring(
+    torch::Tensor index) {
+  const int64_t num_index = index.numel();
+  std::vector<int64_t> shape;
+  shape.push_back(num_index);
+  shape.insert(shape.end(), feature_dim_.begin() + 1, feature_dim_.end());
+  auto result = torch::empty(
+      shape, index.options()
+                 .dtype(dtype_)
+                 .layout(torch::kStrided)
+                 .requires_grad(false));
+  auto future = at::intraop_launch_future([=]() {
+    // The minimum page size to contain one feature.
+    const int64_t aligned_length = (feature_size_ + kDiskAlignmentSize - 1) &
+                                   (long)~(kDiskAlignmentSize - 1);
+
+    const size_t read_buffer_intended_size =
+        (aligned_length + kDiskAlignmentSize) * group_size_ * num_thread_;
+    size_t read_buffer_size =
+        read_buffer_intended_size + kDiskAlignmentSize - 1;
+    auto read_tensor = torch::empty(
+        read_buffer_size,
+        torch::TensorOptions().dtype(torch::kInt8).device(torch::kCPU));
+    auto read_buffer_void_ptr = read_tensor.data_ptr();
+    auto read_buffer = reinterpret_cast<char *>(std::align(
+        kDiskAlignmentSize, read_buffer_intended_size, read_buffer_void_ptr,
+        read_buffer_size));
+    TORCH_CHECK(read_buffer, "read_buffer allocation failed!");
+
+    auto result_buffer = reinterpret_cast<char *>(result.data_ptr());
+
+    // Record the inside offsets of feteched features.
+    auto residual =
+        std::unique_ptr<int64_t[]>(new int64_t[group_size_ * num_thread_]);
+    // Indicator for index error.
+    std::atomic<bool> error_flag{};
+    std::atomic<int64_t> work_queue{};
+    torch::parallel_for(0, num_thread_, 1, [&](int64_t begin, int64_t end) {
+      const auto thread_id = begin;
+      while (true) {
+        begin = work_queue.fetch_add(group_size_, std::memory_order_relaxed);
+        if (begin >= num_index || error_flag.load()) break;
+        end = std::min(begin + group_size_, num_index);
+        int64_t batch_offset = begin;
+        AT_DISPATCH_INDEX_TYPES(
+            index.scalar_type(), "IndexSelectIOUring", ([&] {
+              auto index_data = index.data_ptr<index_t>();
```

Comment:
nit: Is it possible to refactor this to a separate method, to avoid this super deep indent?

Reply:
Reduced indent by 1 by refactoring async logic instead.

Reply:
We can make the code even better as we further visit this part of the code in later PR.